mewarpx/mewarpx/assemblies.py

```diff
@@ -32,6 +48,117 @@ def getvoltage(self):
         """Allows for time-dependent implementations to override this."""
         return mwxrun.eval_expression_t(self.V)
 
+    def getvoltage_e(self):
+        """Allows for time-dependent implementations to override this."""
+        return self.getvoltage() + self.WF
+
+    def init_scrapedparticles(self, fieldlist):
+        """Set up the scraped particles array. Call before
+        append_scrapedparticles.
+
+        Arguments:
+            fieldlist (list): List of string titles for the fields. Order is
+                important; it must match the order for future particle appends
+                that are made.
+        """
+        self._scrapedparticles_fields = fieldlist
+        self._scrapedparticles_data = AppendableArray(
+            typecode='d', unitshape=[len(fieldlist)])
+
+    def record_scrapedparticles(self):
+        """Handles transforming raw particle information from the WarpX
+        scraped particle buffer to the information used to record particles as
+        a function of time.
+
+        Note:
+            Assumes the fixed form of fields given in Assembly().  Doesn't
+            check since this is called many times.
+        """
+
+        # if the conductor doesn't have a label to get scraped particles with
+        # skip it. Currently this will be all conductors except the cathode
+        # and anode.
+        if not hasattr(self, 'scraper_label'):
+            return
+
+        # loop over species and the scraped particle data from the buffer
+        for species in mwxrun.simulation.species:
+            data = np.zeros(7)
+            empty = False
+            idx = None
+
+            # get the number of particles in the buffer - this is primarily
+            # to avoid trying to access the buffer if it is not defined
+            # which causes a segfault
+            buffer_count = _libwarpx.get_num_particles_impacted_boundary(
+                species.name, self.scraper_label
+            )
+            if buffer_count == 0:
+                empty = True
+            else:
+                # get the timesteps at which particles were scraped
+                comp_steps = _libwarpx.get_particle_boundary_buffer(
+                    species.name, self.scraper_label, "stepScraped", mwxrun.lev
+                )
```

Comment:
Won't the step at which particles were scraped always be included when we get the buffer from WarpX? Because the step at which they were scraped will be an attribute of the `ParticleBuffer`.

Reply:
Yes that sounds correct. So do you see something in the code that should be changed due to that fact? I guess the way it is set up we do check that the buffer isn't empty before trying to access it, is that what you are referring to?

Reply:
At the very least there's no reason to pass in the `"stepScraped"` parameter.

Reply:
I don't follow. The `_libwarpx.get_particle_boundary_buffer` function requires a PID name for which you want the buffer.

Reply:
Never mind then